### PR TITLE
[SW-1530] Conversion to H2OFrame needs to work without running H2O client

### PR DESCRIPTION
**Pull request by jakubhava**
First step is to use the init and finalize methods on the remote backend side

### CODE REVIEW

**Comment by jakubhava** [comment]
@mn-mikke good to review, however will be red until we integrate h2o 3.26.0.3 as it has necessary changes

File: core/src/main/scala/org/apache/spark/h2o/backends/external/ExternalWriteConverterCtx.scala [hunk 1]
```diff
@@ -17,54 +17,56 @@
...
+    externalFrameWriter.close()
   }
 
   /**
     * Initialize the communication before the chunks are created
     */
   override def createChunks(keystr: String, expectedTypes: Array[Byte], chunkId: Int, maxVecSizes: Array[Int],
                             sparse: Array[Boolean], vecStartSize: Map[Int, Int]): Unit = {
-    externalFrameWriter.createChunks(keystr, expectedTypes, chunkId, totalNumOfRows, maxVecSizes)
+    externalFrameWriter.createChunk(keystr, expectedTypes, chunkId, totalNumOfRows, maxVecSizes)
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/backends/external/ExternalWriteConverterCtx.scala]: what is a difference between `createChunk` and `createChunks` method?

File: core/src/main/scala/org/apache/spark/h2o/backends/external/ExternalWriteConverterCtx.scala [hunk 2]
```diff
@@ -92,12 +95,34 @@ object ExternalWriteConverterCtx extends ExternalBackendUtils {
...
-      case _ : DataType => ReflectionUtils.supportedTypeOf(dt).javaClass
+      case _: DateType => Long.javaClass
+      case _: UserDefinedType[_ /*ml.linalg.Vector */ ] => classOf[Vector]
+      case _: DataType => ReflectionUtils.supportedTypeOf(dt).javaClass
     }
   }
+
+  def initFrame(key: String, columns: Array[String], timeout: Int): Unit = {
+    val (ip, port, timestamp) = leaderInfo()
+    val client = ExternalFrameWriterClient.create(ip, port, timestamp, timeout)
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/backends/external/ExternalWriteConverterCtx.scala]: Isn't there any overhead with establishing a connection to a leader. Shouldn't we create a connection to a leader only once?

File: core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala [hunk 3]
```diff
@@ -111,3 +116,98 @@ class InternalWriteConverterCtx extends WriteConverterCtx {
...
+    if (!fr.vecs().isEmpty) {
+      val layouts = fr.vecs().map(_.espc())
+      val first = layouts.head
+      layouts.tail.foreach { espc =>
+        if (first != espc) {
+          throw new IllegalArgumentException(
+            s"""
+               | Invalid shape of H2O Frame:
+               |
+          | ${layouts.zipWithIndex.map { case (arr, idx) => s"Vec $idx has layout: ${arr.mkString(", ")}\n" }}
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala]: nit: indentation

File: core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala [hunk 4]
```diff
@@ -111,3 +116,98 @@ class InternalWriteConverterCtx extends WriteConverterCtx {
...
+    }
+
+    // Validate num of chunks in each vector
+    if (!fr.vecs().isEmpty) {
+      val first = fr.vecs.head
+      fr.vecs.tail.foreach { vec =>
+        if (vec.nChunks() != first.nChunks()) {
+          throw new IllegalArgumentException(
+            s"""
+               | Vectors have different number of chunks: ${fr.vecs().map(_.nChunks()).mkString(", ")}
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala]: nit: one line?

File: core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala [hunk 5]
```diff
@@ -111,3 +116,98 @@ class InternalWriteConverterCtx extends WriteConverterCtx {
...
+        nodes.foreach {
+          case (node, n) =>
+            Log.debug(node + ": " + n + " chunks.")
+        }
+      }
+    }
+
+    // Validate num of chunks in each vector
+    if (!fr.vecs().isEmpty) {
+      val first = fr.vecs.head
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala]: Would it be possible to extract validation logic to a separated method? [MRTask](https://github.com/h2oai/sparkling-water/pull/1451/files#diff-21fbbe6ed7d8cdb3212cdcd8dc0c2bceR193) seems to use something similar.

File: core/src/main/scala/org/apache/spark/h2o/backends/external/ExternalWriteConverterCtx.scala (hunk 1 above)
Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/backends/external/ExternalWriteConverterCtx.scala]: They are the same, we changed the name of the method on the H2O backend and this just just updates the code according to that

File: core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala (hunk 4 above)
Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala]: 👍

File: core/src/main/scala/org/apache/spark/h2o/backends/external/ExternalWriteConverterCtx.scala (hunk 2 above)
Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/backends/external/ExternalWriteConverterCtx.scala]: Good point!

File: core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala [hunk 6]
```diff
@@ -111,3 +128,63 @@ class InternalWriteConverterCtx extends WriteConverterCtx {
     (vector.size until maxVecSize).foreach(idx => put(startIdx + idx, 0.0))
   }
 }
+
+object InternalWriteConverterCtx {
+
+  private def validateFrame(fr: Frame): Unit = {
+
+
```
Comment (mn-mikke) [inline core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala]: nit: extra lines

Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala]: thanks!

File: core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala (hunk 5 above)
Comment (jakubhava) [reply core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalWriteConverterCtx.scala]: I have tried that, it creates slightly shorter code, but hurts readability.

**Pull request merged**